python/cudf/cudf/core/buffer.py

```diff
@@ -5,108 +5,206 @@
+        integer.
     owner : object, optional
-        Python object to which the lifetime of the memory
-        allocation is tied. If provided, a reference to this
-        object is kept in this Buffer.
+        Python object to which the lifetime of the memory allocation is tied.
+        A reference to this object is kept in the returned Buffer.
     """
 
     _ptr: int
     _size: int
     _owner: object
 
     def __init__(
-        self, data: Any = None, size: int = None, owner: object = None
+        self, data: Union[int, Any], *, size: int = None, owner: object = None
     ):
-        if isinstance(data, Buffer):
-            self._ptr = data._ptr
-            self._size = data.size
-            self._owner = owner or data._owner
-        elif isinstance(data, rmm.DeviceBuffer):
-            self._ptr = data.ptr
-            self._size = data.size
-            self._owner = data
-        elif hasattr(data, "__array_interface__") or hasattr(
-            data, "__cuda_array_interface__"
-        ):
-            self._init_from_array_like(data, owner)
-        elif isinstance(data, memoryview):
-            self._init_from_array_like(np.asarray(data), owner)
-        elif isinstance(data, int):
-            if not isinstance(size, int):
-                raise TypeError("size must be integer")
+        if isinstance(data, int):
+            if size is None:
+                raise ValueError(
+                    "size must be specified when `data` is an integer"
+                )
             self._ptr = data
             self._size = size
             self._owner = owner
-        elif data is None:
-            self._ptr = 0
-            self._size = 0
-            self._owner = None
         else:
-            try:
-                data = memoryview(data)
-            except TypeError:
-                raise TypeError("data must be Buffer, array-like or integer")
-            self._init_from_array_like(np.asarray(data), owner)
-
-    @classmethod
-    def from_buffer(cls, buffer: Buffer, size: int = None, offset: int = 0):
-        """
-        Create a buffer from another buffer
-
-        Parameters
-        ----------
-        buffer : Buffer
-            The base buffer, which will also be set as the owner of
-            the memory allocation.
-        size : int, optional
-            Size of the memory allocation (default: `buffer.size`).
-        offset : int, optional
-            Start offset relative to `buffer.ptr`.
-        """
+            if size is not None or owner is not None:
+                raise ValueError(
+                    "`size` and `owner` must be None when "
+                    "`data` is an buffer-like object"
+                )
+
+            # `data` is a buffer-like object
+            buf: Any = data
+            if isinstance(buf, rmm.DeviceBuffer):
+                self._ptr, self._size, self._owner = buf.ptr, buf.size, buf
```

Comment:
Tuple unpacking can be helpful when there are multiple return values but I would avoid it for packing/unpacking multiple assignments.
```suggestion
                self._ptr = buf.ptr
                self._size = buf.size
                self._owner = buf
```